File: apps/worker/src/summary/format.ts
/**
 * Parsing and formatting of the mandatory 5-section Russian summary structure.
 *
 * Expected Claude output format:
 *   Заголовок: <headline>
 *   Что произошло: <1–2 sentences>
 *   Почему важно: <1 sentence>
 *   Что дальше: <1 sentence>
 *   Источники: <source names>
 */

export interface ParsedSummary {
  title: string;
  whatHappened: string;
  whyImportant: string;
  whatsNext: string;
  sources: string;
}

const SECTIONS = [
  { key: 'Заголовок', field: 'title' },
  { key: 'Что произошло', field: 'whatHappened' },
  { key: 'Почему важно', field: 'whyImportant' },
  { key: 'Что дальше', field: 'whatsNext' },
  { key: 'Источники', field: 'sources' },
] as const;

/**
 * Parse Claude output into structured sections.
 * Each section starts with "Key: value" on its own line.
 * Multi-line continuation is supported for section content.
 * Returns null if any required section is missing or has an empty value.
 */
export function parseSections(text: string): ParsedSummary | null {
  const lines = text.split('\n');
  const result: Partial<ParsedSummary> = {};

  for (let sIdx = 0; sIdx < SECTIONS.length; sIdx++) {
    const { key, field } = SECTIONS[sIdx]!;
    const nextKey = sIdx + 1 < SECTIONS.length ? SECTIONS[sIdx + 1]!.key : null;

    // Find the line that begins this section
    const startIdx = lines.findIndex(l => l.trim().startsWith(key + ':'));
    if (startIdx === -1) return null;

    // Find where this section ends (start of next section or EOF)
    let endIdx = lines.length;
    if (nextKey !== null) {
      const ni = lines.findIndex((l, i) => i > startIdx && l.trim().startsWith(nextKey + ':'));
      if (ni !== -1) endIdx = ni;
    }

    // Collect the value: inline text after "Key:" + any continuation lines
    const firstLine = lines[startIdx]!.trim().slice(key.length + 1).trim();
    const continuations = lines
      .slice(startIdx + 1, endIdx)
      .map(l => l.trim())
      .filter(Boolean)
      .join(' ');

    const value = continuations ? `${firstLine} ${continuations}`.trim() : firstLine;
    if (!value) return null;

    (result as Record<string, string>)[field] = value;
  }

  return result as ParsedSummary;
}

/**
 * The body text used for character-length checks (excludes the Источники line).
 */
export function formatBody(parsed: ParsedSummary): string {
  return [
    `Что произошло: ${parsed.whatHappened}`,
    `Почему важно: ${parsed.whyImportant}`,
    `Что дальше: ${parsed.whatsNext}`,
  ].join('\n');
}

/**
 * Full display text stored in summary_ru (title + body + sources).
 */
export function formatFull(parsed: ParsedSummary): string {
  return [
    parsed.title,
    '',
    `Что произошло: ${parsed.whatHappened}`,
    `Почему важно: ${parsed.whyImportant}`,
    `Что дальше: ${parsed.whatsNext}`,
    `Источники: ${parsed.sources}`,
  ].join('\n');
}

File: apps/local-engine/summary/format.py
"""summary/format.py — Parse and format the mandatory 5-section Russian summary.

Exact port of apps/worker/src/summary/format.ts.

Expected LLM output format:
  Заголовок: <headline>
  Что произошло: <1–2 sentences>
  Почему важно: <1 sentence>
  Что дальше: <1 sentence>
  Источники: <source names>
"""

from __future__ import annotations

from dataclasses import dataclass


@dataclass
class ParsedSummary:
    title: str
    what_happened: str
    why_important: str
    whats_next: str
    sources: str


# Section keys in mandatory order — identical to TS SECTIONS array.
_SECTIONS = [
    ("Заголовок", "title"),
    ("Что произошло", "what_happened"),
    ("Почему важно", "why_important"),
    ("Что дальше", "whats_next"),
    ("Источники", "sources"),
]


def parse_sections(text: str) -> ParsedSummary | None:
    """Parse LLM output into structured sections.

    Each section starts with "Key: value" on its own line.
    Multi-line continuation is supported for section content.
    Returns None if any required section is missing or has an empty value.
    """
    lines = text.split("\n")
    result: dict[str, str] = {}

    for s_idx, (key, field_name) in enumerate(_SECTIONS):
        next_key = _SECTIONS[s_idx + 1][0] if s_idx + 1 < len(_SECTIONS) else None

        # Find the line that begins this section
        start_idx = next(
            (i for i, line in enumerate(lines) if line.strip().startswith(key + ":")),
            -1,
        )
        if start_idx == -1:
            return None

        # Find where this section ends (start of next section or EOF)
        end_idx = len(lines)
        if next_key is not None:
            ni = next(
                (
                    i
                    for i, line in enumerate(lines)
                    if i > start_idx and line.strip().startswith(next_key + ":")
                ),
                -1,
            )
            if ni != -1:
                end_idx = ni

        # Collect value: inline text after "Key:" + any continuation lines
        first_line = lines[start_idx].strip()[len(key) + 1 :].strip()
        continuations = " ".join(
            line.strip()
            for line in lines[start_idx + 1 : end_idx]
            if line.strip()
        )
        value = (f"{first_line} {continuations}".strip() if continuations else first_line)
        if not value:
            return None

        result[field_name] = value

    return ParsedSummary(
        title=result["title"],
        what_happened=result["what_happened"],
        why_important=result["why_important"],
        whats_next=result["whats_next"],
        sources=result["sources"],
    )


def format_body(parsed: ParsedSummary) -> str:
    """Body text used for character-length checks (excludes Источники and title)."""
    return "\n".join(
        [
            f"Что произошло: {parsed.what_happened}",
            f"Почему важно: {parsed.why_important}",
            f"Что дальше: {parsed.whats_next}",
        ]
    )


def format_full(parsed: ParsedSummary) -> str:
    """Full display text stored in summary_ru (title + body + sources)."""
    return "\n".join(
        [
            parsed.title,
            "",
            f"Что произошло: {parsed.what_happened}",
            f"Почему важно: {parsed.why_important}",
            f"Что дальше: {parsed.whats_next}",
            f"Источники: {parsed.sources}",
        ]
    )
